### DISCUSSION TITLE
Reduce the API scope

### DISCUSSION THREAD

**Discussion by ramsey** [php-identifier/identifier#8] [Ideas]
@ocramius writes on Twitter:[^1]

> I would turn down the API surface a notch: you are repeating the "I want to support every use-case" mistake you were going for before.

I think this is a fair criticism, and looking at what I've got here, it does look like I'm trying to boil the ocean.

So, what's the bare minimum?

Some things that come to mind right away are getting rid of the specific interfaces for UUIDs and ULIDs and perhaps introducing a more generic "time-based" identifier interface that only provides `getDateTime()`.

What else?

[^1]: https://twitter.com/Ocramius/status/1576364216889376768

**Comment by Ocramius** [comment]
I would say that it's pretty much OK for ULID/UUID to have more methods, but that's not needed for a common implementation.

When I implemented `LazyUuidFromString` in `ramsey/uuid`, the amount of API surface to cover was enormous.

From a consumer PoV, a string/integer/binary UUID is used for:

 * serialization (`toString()`, `toInt()`, `toBytes()`)
 * comparison (`equals()`)

Not much else going on there that should be interfaced, I suppose.

Everything else should be a "nice to have" that can be useful to 1 in 100 users, but can then have a separate API dealing just with those edge cases.

**Comment by ramsey** [comment]
I greatly simplified the API and significantly reduced scope in https://github.com/php-identifier/identifier/commit/90a68eb054b17269df74f6329225edd12904a9bf

**Reply by MGatner** [reply]
Much cleaner! I'm a little sad to lose the explicit path to a ULID implementation, but give  that all the same pieces are still there it just means a bit of composition on the developer. I went and thought about how I would use this with Symfony\Ulid and it seems reasonable. 

Thanks for keeping us updated! 🙇‍♂️

**Reply by Ocramius** [reply]
Much simpler, well done!
